src/main/java/org/spongepowered/common/network/channel/SpongeChannelRegistry.java

```diff
@@ -24,29 +24,298 @@
+        this.registerChannelType(PacketChannel.class, SpongePacketChannel::new);
+    }
+
+    public ChannelBufferAllocator getBufferAllocator() {
+        return this.bufferAllocator;
+    }
+
+    private <T extends Channel> void registerChannelType(
+            final Class<T> channelType, final BiFunction<ResourceKey, SpongeChannelRegistry, T> builder) {
+        this.channelBuilders.put(channelType, (BiFunction<ResourceKey, SpongeChannelRegistry, Channel>) builder);
+    }
+
+    public <C extends Channel> C createChannel(final ResourceKey channelKey, final Class<C> channelType) throws DuplicateRegistrationException {
+        requireNonNull(channelKey, "channelKey");
+        requireNonNull(channelType, "channelType");
+        if (this.channels.containsKey(channelKey)) {
+            throw new DuplicateRegistrationException("The channel key \"" + channelKey + "\" is already in use.");
+        }
+        final @Nullable BiFunction<ResourceKey, SpongeChannelRegistry, Channel> builder = this.channelBuilders.get(channelType);
+        if (builder == null) {
+            throw new IllegalArgumentException("Unsupported channel type: " + channelType);
+        }
+        final Channel channel = builder.apply(channelKey, this);
+        this.channels.put(channelKey, channel);
+        return (C) channel;
+    }
 
     @Override
-    public Optional<Channel> get(ResourceKey channelKey) {
-        return Optional.empty();
+    public Optional<Channel> get(final ResourceKey channelKey) {
+        requireNonNull(channelKey, "channelKey");
+        return Optional.ofNullable(this.channels.get(channelKey));
     }
 
     @Override
-    public <C extends Channel> C getOfType(ResourceKey channelKey, Class<C> channelType) {
-        return null;
+    public <C extends Channel> C getOfType(final ResourceKey channelKey, final Class<C> channelType) {
+        requireNonNull(channelKey, "channelKey");
+        requireNonNull(channelType, "channelType");
+        final @Nullable Channel binding = this.channels.get(channelKey);
+        if (binding != null) {
+            if (!channelType.isInstance(binding)) {
+                throw new IllegalStateException("There's already a channel registered for "
+                        + channelKey + ", but it is not of the requested type " + channelType);
+            }
+            return (C) binding;
+        }
+        return this.createChannel(channelKey, channelType);
     }
 
     @Override
     public Collection<Channel> getChannels() {
-        return null;
+        return ImmutableList.copyOf(this.channels.values());
+    }
+
+    private static final class ChannelRegistrationsResult {
+
+        private final CompletableFuture<Void> future;
+
+        private ChannelRegistrationsResult(final CompletableFuture<Void> future) {
+            this.future = future;
+        }
+    }
+
+    public void postRegistryEvent() {
+        final Cause cause = Cause.of(EventContext.empty(), this);
```

Comment:
Can use the PhaseTracker as is now. But if this whole class shebang is usable from multiple threads, you may have to change the map used to at least ConcurrentHashMap.

Reply:
Depends where you call this in the lifecycle.

Part of the lifecycle is before the engine(s) are even created. If so, cannot use PhaseTracker.

Even if you could, don't see a point. We explicitly know what the cause will always be here, no reason to grab one from the CauseStack.